lib/metasploit/framework/login_scanner/teamcity.rb

```diff
@@ -0,0 +1,181 @@
+            sleep(timeout + 1) # + 1 as TeamCity is off-by-one when reporting the lockout timer.
+            result = try_login(username, password, public_key)
+            return result
+          end
+
+          return { status: INCORRECT, proof: res } if res.body.match?('Incorrect username or password')
+          return { status: UNABLE_TO_CONNECT, proof: res } if res.body.match?('ajax') # TODO: Get the exact error message here.
+          return { status: INVALID_PUBLIC_PART, proof: res } if res.body.match?('publicKeyExpired') # TODO: Invalid public part? Or Incorrect/Unable_to_connect?
+
+          { status: :success, proof: res }
+        end
+
+        # Send a logout request for the provided user's headers.
+        # This header stores the user's cookie.
+        # @return [Hash] A hash with the status and an error or the response.
+        def logout(headers)
+          logout_params = {
+            'method' => 'POST',
+            'uri' => normalize_uri(@uri.to_s, LOGOUT_PAGE),
+            'headers' => headers
+          }
+
+          begin
+            logout_res = send_request(logout_params)
+          rescue ::Rex::ConnectionError, ::Rex::ConnectionProxyError, ::Errno::ECONNRESET, ::Errno::EINTR, ::Rex::TimeoutError, ::Timeout::Error, ::EOFError => e
+            return { status: UNABLE_TO_CONNECT, proof: e }
+          end
+
+          return { status: UNABLE_TO_CONNECT, proof: 'Unable to connect to the TeamCity service' } if logout_res.nil?
+          # A successful logout request wants to redirect us back to the login page
+          return { status: UNABLE_TO_CONNECT, proof: "Received an unexpected status code: #{logout_res.code}" } if logout_res.code != 302
+
+          { status: :success, proof: logout_res }
+        end
+
+        def attempt_login(credential)
+          result_options = {
+            credential:   credential,
+            host:         @host,
+            port:         @port,
+            protocol:     'tcp',
+            service_name: 'teamcity'
+          }
+
+          # Needed to retrieve the public key that will be used to encrypt the user's password.
+          page_data = get_page_data
+          return Result.new(result_options.merge(page_data)) if page_data[:status] != :success
+
+          public_key_result = get_public_key(page_data[:proof])
+          return Result.new(result_options.merge(public_key_result)) if public_key_result[:status] != :success
+
+          login_result = try_login(credential.public, credential.private, public_key_result[:proof])
+          return Result.new(result_options.merge(login_result)) if login_result[:status] != :success
+
+          # Ensure we log the user out, so that our logged in session does not appear under the user's profile.
+          logout_result = logout(login_result[:proof].headers)
+          return Result.new(result_options.merge(logout_result)) if logout_result[:status] != :success
```

Comment:
I think this should probably be fire and forget, or a separate warning to the user, or we keep the original status code and additionally supplement data with the fact they didn't have their session cleaned up correctly. But we definitely shouldn't treat a logout network failure as an overall cred failure